### PR TITLE
[UX] A new look of SkyPilot console outputs

### PR DESCRIPTION
**Pull request by Michaelvll**
This PR revamped the output of SkyPilot console output.

## Launch
image

## Managed Jobs
image

image

## Service
image

TODO:
- [x] Test logging for jobs and serve
- [ ] Use detach run by default for `jobs launch` for more responsive feeling as serve does (for discussion; can be done in a future PR)

Tested (run the relevant ones):

- [x] Code formatting: `bash format.sh`
- [x] Any manual or new tests for this PR (please specify below)
  - [x] Clouds with old provisioner
- [x] All smoke tests: `pytest tests/test_smoke.py` (except for `test_tpu_vm_pod` due to quota issue, `TestStorageWithCredentials::test_aws_regions[eu-west-3]` (and other 8 regions) due to permission issue).
- [ ] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_fill_in_the_name` 
- [x] Backward compatibility tests: `conda deactivate; bash -i tests/backward_compatibility_tests.sh`

### CODE REVIEW

**Comment by Michaelvll** [comment]
The smoke test is now passed. PTAL.

File: sky/provision/azure/config.py [hunk 1]
```diff
@@ -92,10 +94,19 @@ def bootstrap_instances(
                 retry += 1
                 continue
             raise
+        except azure.exceptions().ClientAuthenticationError as e:
+            message = (
+                'Failed to authenticate with Azure. Please check your Azure '
+                f'credentials. Error: {common_utils.format_exception(e)}'
+            ).replace('\n', ' ')
+            logger.error(message)
+            raise exceptions.NoClusterLaunchedError(message) from e
```
Comment (concretevitamin) [inline sky/provision/azure/config.py:103]: Q: why do we need a new exception?

File: sky/backends/cloud_vm_ray_backend.py [hunk 2]
```diff
@@ -1566,9 +1595,9 @@ def _retry_zones(
                     self._ensure_cluster_ray_started(handle, log_abs_path)
 
                 config_dict['handle'] = handle
-                plural = '' if num_nodes == 1 else 's'
-                logger.info(f'{fore.GREEN}Successfully provisioned or found'
-                            f' existing VM{plural}.{style.RESET_ALL}')
+                logger.info(
+                    ux_utils.finishing_message(
+                        f'Cluster {cluster_name!r} launched.', log_path))
```
Comment (concretevitamin) [inline sky/backends/cloud_vm_ray_backend.py:1600]: Hmm, when is this shown? Seems to always show `Cluster launched: xxx`

File: sky/utils/rich_utils.py [hunk 3]
```diff
@@ -30,19 +31,63 @@ def start(self):
...
-            _status = console.status(msg)
-        _status.update(msg)
-        return _status
+            _status = console.status(msg, refresh_per_second=4)
+        return NestedStatus(msg)
     return _NoOpConsoleStatus()
 
 
+def stop_safe_status():
+    """Stop the safe status."""
```
Comment (concretevitamin) [inline sky/utils/rich_utils.py:85]: Give some guidance to callers: when to / why call this?

File: sky/exceptions.py [hunk 4]
```diff
@@ -291,3 +291,8 @@ class PortDoesNotExistError(Exception):
 class UserRequestRejectedByPolicy(Exception):
     """Raised when a user request is rejected by an admin policy."""
     pass
+
+
+class NoClusterLaunchedError(Exception):
```
Comment (concretevitamin) [inline sky/exceptions.py:296]: Why needed?

File: sky/provision/provisioner.py [hunk 5]
```diff
@@ -42,76 +43,50 @@
...
-                    f'{region_name}{style.RESET_ALL} ({zone_str})')
-
     start = time.time()
-    with rich_utils.safe_status('[bold cyan]Launching[/]') as status:
+    # TODO(suquark): Should we cache the bootstrapped result?
+    #  Currently it is not necessary as bootstrapping takes
+    #  only ~3s, caching it seems over-engineering and could
+    #  cause other issues like the cache is not synced
+    #  with the cloud configuration.
+    config = provision.bootstrap_instances(provider_name, region_name,
```
Comment (concretevitamin) [inline sky/provision/provisioner.py:58]: Why this change? Looks like we don't catch exceptions as we did before.

**Review by concretevitamin — COMMENTED** [review]
Very nice @Michaelvll. Did a pass.

**Comment by romilbhardwaj** [comment]
This is very cool!

A quick nit: can we add a space after the gear emoji? Also is the space after the useful commands intended?

`⚙️ Launching on GCP us-central1 (us-central1-a).`

image


Commit: `d3965aa1ee39ded4478b4475b7d78ccc1f0f3513`

File: sky/utils/ux_utils.py [hunk 6]
```diff
@@ -102,3 +111,50 @@ def run(self, *args, **kwargs):
                 with ux_utils.enable_traceback():
                     logger.error(f'  Traceback:\n{traceback.format_exc()}')
                 raise
+
+
+def starting_message(message: str) -> str:
```
[the remark this answers is not in the document]
Comment (Michaelvll) [reply sky/utils/ux_utils.py:116]: Yes, we switched all the significant messages to use these functions.

File: sky/skylet/providers/lambda_cloud/node_provider.py [hunk 7]
```diff
@@ -25,7 +25,7 @@
 _REMOTE_SSH_KEY_NAME = '~/.lambda_cloud/ssh_key_name'
 _REMOTE_RAY_SSH_KEY = '~/ray_bootstrap_key.pem'
 _REMOTE_RAY_YAML = '~/ray_bootstrap_config.yaml'
-_GET_INTERNAL_IP_CMD = 'ip -4 -br addr show | grep UP | grep -Eo "(10\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)|172\.(1[6-9]|2[0-9]|3[0-1]))\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)"'
+_GET_INTERNAL_IP_CMD = 's=$(ip -4 -br addr show | grep UP); echo "$s"; echo "$s" | grep -Eo "(10\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)|172\.(1[6-9]|2[0-9]|3[0-1])|104\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?))\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)"'
```
[the remark this answers is not in the document]
Comment (Michaelvll) [reply sky/skylet/providers/lambda_cloud/node_provider.py:28]: This is discovered during the test with lambda. It might be fine to just have it in. Created another PR for this: https://github.com/skypilot-org/skypilot/pull/4072

File: sky/backends/cloud_vm_ray_backend.py (hunk 2 above)
Comment (Michaelvll) [reply sky/backends/cloud_vm_ray_backend.py:1600]: Ahh, this is for legacy clouds using the old code path (node_provider/ray autoscaler) to provisioned. Changed the message to be consistent.

File: sky/exceptions.py (hunk 4 above)
Comment (Michaelvll) [reply sky/exceptions.py:296]: This do the optimization we did for previous ray autoscaler-based implementation, i.e. we should not need to invoke the cluster termination if the provisioner never created any resources, e.g., no VPC is found, etc.

File: sky/skylet/providers/lambda_cloud/node_provider.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (concretevitamin) [reply sky/skylet/providers/lambda_cloud/node_provider.py:28]: Ah no problem. Just curious.

File: sky/provision/provisioner.py (hunk 5 above)
Comment (Michaelvll) [reply sky/provision/provisioner.py:58]: Our original handling is just catching the exception, log it and raise again. We don't need that logging as we should expect the cloud-specific implementation should log them correctly, and the original region, cloud information is duplicated with the existing logs.

File: sky/provision/azure/config.py (hunk 1 above)
Comment (Michaelvll) [reply sky/provision/azure/config.py:103]: See https://github.com/skypilot-org/skypilot/pull/4023#discussion_r1797411690

File: sky/exceptions.py (hunk 4 above)
Comment (Michaelvll) [reply sky/exceptions.py:296]: Added a docstr for this error.

**Comment by Michaelvll** [comment]
> Also is the space after the useful commands intended?

This new line was intended, as we have observed before that sometimes, if people ctrl-c or the program exited abnormally, users' terminal line will be shown in the same line as the last output, which looks messy. We added that newline to prevent those cases. Wdyt?

**Review by concretevitamin — APPROVED** [review]
